File: server/app/models/financial_account.py
"""Financial account model for storing user's financial accounts.

This model has been migrated to SQLAlchemy 2.0 with Mapped[...] annotations.
"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import TYPE_CHECKING, Optional
from uuid import UUID, uuid4 as uuid4_factory

from sqlalchemy import Boolean, CheckConstraint, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, col

if TYPE_CHECKING:
    from app.models.user import User


class AccountNature(str, Enum):
    """Account nature enum - defines whether account is an asset or liability."""

    ASSET = "ASSET"
    LIABILITY = "LIABILITY"


class AccountType(str, Enum):
    """Account type enum - specific account categories."""

    CASH = "CASH"
    DEPOSIT = "DEPOSIT"
    E_MONEY = "E_MONEY"
    INVESTMENT = "INVESTMENT"
    RECEIVABLE = "RECEIVABLE"
    CREDIT_CARD = "CREDIT_CARD"
    LOAN = "LOAN"
    PAYABLE = "PAYABLE"


class AccountStatus(str, Enum):
    """Account status enum."""

    ACTIVE = "ACTIVE"
    INACTIVE = "INACTIVE"
    CLOSED = "CLOSED"


class FinancialAccount(Base):
    """Financial account model for storing user's financial accounts.

    This model represents various financial accounts like bank accounts,
    credit cards, cash wallets, etc.

    Attributes:
        id: Primary key (UUID)
        user_uuid: Foreign key to users table (UUID)
        name: Account name (e.g., "招商银行储蓄卡")
        nature: Account nature ('ASSET' or 'LIABILITY')
        type: Account type (e.g., 'CASH', 'DEPOSIT', 'CREDIT_CARD')
        currency_code: Currency code (default: 'CNY')
        initial_balance: Initial account balance (precision: 20,8)
        current_balance: Current account balance (precision: 20,8)
        include_in_net_worth: Whether to include in net worth calculation
        include_in_cash_flow: Whether to include in daily cash flow forecast
        status: Account status ('ACTIVE', 'INACTIVE', 'CLOSED')
        created_at: Record creation timestamp
        updated_at: Record update timestamp
        user: Relationship to user
    """

    __tablename__ = "financial_accounts"
    __table_args__ = (
        CheckConstraint("nature IN ('ASSET', 'LIABILITY')", name="check_nature"),
        CheckConstraint("status IN ('ACTIVE', 'INACTIVE', 'CLOSED')", name="check_status"),
    )

    id: Mapped[UUID] = col.uuid_pk(uuid4_factory)
    user_uuid: Mapped[UUID] = col.uuid_fk("users", ondelete="CASCADE", index=True, column="uuid")
    name: Mapped[str] = mapped_column(String(100))
    nature: Mapped[str] = mapped_column(String(20))
    type: Mapped[str | None] = mapped_column(String(50), nullable=True)
    currency_code: Mapped[str] = mapped_column(String(3), default="CNY")
    initial_balance: Mapped[Decimal] = col.numeric(precision=20, scale=8, default=Decimal("0"))
    current_balance: Mapped[Decimal] = col.numeric(precision=20, scale=8, default=Decimal("0"))
    include_in_net_worth: Mapped[bool] = mapped_column(Boolean, default=True)
    include_in_cash_flow: Mapped[bool] = mapped_column(Boolean, default=False)
    status: Mapped[str] = mapped_column(String(20), default="ACTIVE")
    created_at: Mapped[datetime] = col.timestamptz()
    updated_at: Mapped[datetime | None] = col.timestamptz(nullable=True)

    user: Mapped[User | None] = relationship(
        "User",
        back_populates="financial_accounts",
        foreign_keys="[FinancialAccount.user_uuid]",
        primaryjoin="FinancialAccount.user_uuid == User.uuid",
    )

    @property
    def balance_float(self) -> float:
        """Get initial balance as float."""
        return float(self.initial_balance)

    @property
    def is_asset(self) -> bool:
        """Check if account is an asset."""
        return self.nature == AccountNature.ASSET.value

    @property
    def is_liability(self) -> bool:
        """Check if account is a liability."""
        return self.nature == AccountNature.LIABILITY.value

    @property
    def is_active(self) -> bool:
        """Check if account is active."""
        return self.status == AccountStatus.ACTIVE.value

    @property
    def net_worth_contribution(self) -> Decimal:
        """Calculate this account's contribution to net worth."""
        if not self.include_in_net_worth or not self.is_active:
            return Decimal("0")
        if self.is_liability:
            return -abs(self.initial_balance)
        return self.initial_balance

File: client/lib/features/profile/models/financial_account.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:decimal/decimal.dart';

part 'financial_account.freezed.dart';
part 'financial_account.g.dart';

/// Account display info for cross-currency views
@freezed
abstract class AccountDisplay with _$AccountDisplay {
  const factory AccountDisplay({
    required String sign,
    required String value,
    required String valueFormatted,
    required String currencySymbol,
    required String fullString,
  }) = _AccountDisplay;

  factory AccountDisplay.fromJson(Map<String, dynamic> json) =>
      _$AccountDisplayFromJson(json);
}

/// Financial nature enum - Corresponds to the nature field in backend API
enum FinancialNature {
  @JsonValue('ASSET')
  asset,
  @JsonValue('LIABILITY')
  liability,
}

/// Financial account type enum - Corresponds to the type field in backend API
enum FinancialAccountType {
  @JsonValue('CASH')
  cash,
  @JsonValue('DEPOSIT')
  deposit,
  @JsonValue('E_MONEY')
  eMoney,
  @JsonValue('INVESTMENT')
  investment,
  @JsonValue('RECEIVABLE')
  receivable,
  @JsonValue('CREDIT_CARD')
  creditCard,
  @JsonValue('LOAN')
  loan,
  @JsonValue('PAYABLE')
  payable,
}

/// Account status enum - Corresponds to the status field in backend API
enum AccountStatus {
  @JsonValue('ACTIVE')
  active,
  @JsonValue('INACTIVE')
  inactive,
  @JsonValue('CLOSED')
  closed,
}

/// Financial account data model - Corresponds to backend API response
@freezed
abstract class FinancialAccount with _$FinancialAccount {
  const factory FinancialAccount({
    /// Account ID (UUID from backend)
    String? id,

    /// Account name
    required String name,

    /// Account nature: ASSET or LIABILITY
    required FinancialNature nature,

    /// Account type: CASH, DEPOSIT, E_MONEY etc.
    FinancialAccountType? type,

    /// Currency code (Default: CNY)
    @Default('CNY') String currencyCode,

    /// Initial balance
    @JsonKey(fromJson: _decimalFromJson, toJson: _decimalToJson)
    required Decimal initialBalance,

    /// Current balance
    @JsonKey(fromJson: _decimalFromJsonNullable, toJson: _decimalToJsonOrZero)
    Decimal? currentBalance,

    /// Whether to include in net worth
    @Default(true) bool includeInNetWorth,

    /// Whether to include in daily cash flow forecast (Liquidity tag)
    @Default(false) bool includeInCashFlow,

    /// Display info (optional, used for cross-currency summary display)
    AccountDisplay? display,

    /// Account status
    @Default(AccountStatus.active) AccountStatus status,

    /// Creation time (ISO 8601 string)
    String? createdAt,

    /// Update time (ISO 8601 string)
    String? updatedAt,
  }) = _FinancialAccount;

  factory FinancialAccount.fromJson(Map<String, dynamic> json) =>
      _$FinancialAccountFromJson(json);
}

/// Financial account summary response model
@freezed
abstract class FinancialAccountSummary with _$FinancialAccountSummary {
  const factory FinancialAccountSummary({
    @JsonKey(fromJson: _decimalFromJson, toJson: _decimalToJson)
    required Decimal totalBalance,
    required DateTime lastUpdatedAt,
  }) = _FinancialAccountSummary;

  factory FinancialAccountSummary.fromJson(Map<String, dynamic> json) =>
      _$FinancialAccountSummaryFromJson(json);
}

/// Financial account list response model
@freezed
abstract class FinancialAccountResponse with _$FinancialAccountResponse {
  const factory FinancialAccountResponse({
    @Default([]) List<FinancialAccount> accounts,
    @JsonKey(fromJson: _decimalFromJsonNullable, toJson: _decimalToJson)
    required Decimal totalBalance,
    @Default('') String lastUpdatedAt,
  }) = _FinancialAccountResponse;

  factory FinancialAccountResponse.fromJson(Map<String, dynamic> json) =>
      _$FinancialAccountResponseFromJson(json);
}

/// Financial account request model
@freezed
abstract class FinancialAccountRequest with _$FinancialAccountRequest {
  @JsonSerializable(explicitToJson: true)
  const factory FinancialAccountRequest({
    required List<FinancialAccount> accounts,
  }) = _FinancialAccountRequest;

  factory FinancialAccountRequest.fromJson(Map<String, dynamic> json) =>
      _$FinancialAccountRequestFromJson(json);
}

// Helper functions: Decimal serialization/deserialization
Decimal _decimalFromJson(dynamic value) {
  if (value is String) return Decimal.parse(value);
  if (value is num) return Decimal.parse(value.toString());
  return Decimal.zero;
}

Decimal _decimalFromJsonNullable(dynamic value) {
  if (value == null) return Decimal.zero;
  return _decimalFromJson(value);
}

String _decimalToJson(Decimal value) => value.toString();
String _decimalToJsonOrZero(Decimal? value) => value?.toString() ?? '0';
